modules/tensor_mechanics/include/kernels/MomentBalancing.h

```diff
@@ -27,18 +37,29 @@ class MomentBalancing : public Kernel
   virtual Real computeQpJacobian();
   virtual Real computeQpOffDiagJacobian(unsigned int jvar);
 
+  /// the stress tensor (not the moment stress) at the quad-point.
   const MaterialProperty<RankTwoTensor> & _stress;
+
+  /**
+   * d(stress tensor)/(d strain tensor)
+   * Here strain_ij = grad_j disp_i + epsilon_ijk * wc_k
+   */
   const MaterialProperty<RankFourTensor> & _Jacobian_mult;
 
-private:
+  /// The Kernel computes epsilon_{component j k}*stress_{j k}
   const unsigned int _component;
 
-  const unsigned int _wc_x_var;
-  const unsigned int _wc_y_var;
-  const unsigned int _wc_z_var;
-  const unsigned int _xdisp_var;
-  const unsigned int _ydisp_var;
-  const unsigned int _zdisp_var;
+  // number of Cosserat rotation components provided by the user
+  const unsigned int _nrots;
+
+  /// the moose variable numbers for the Cosserat rotation degrees of freedom
+  std::vector<unsigned int> _wc_var;
+
+  // number of displacement components provided by the user
+  const unsigned int _ndisp;
```

Comment:
At the moment, ndisp and nrots must both equal 3.  However, if the crowds of users clamor for a 2D version then ndisp=2 and nrots=1.